packages/experiments/src/utilities/BaseState.tsx

```diff
@@ -1,86 +1,57 @@
-import { IStateComponentProps } from '../Foundation';
-import { BaseComponent } from '../Utilities';
+import * as React from 'react';
 
-// TODO: add transforms and flesh out typing
-export type ITransform<TComponentProps, TViewProps> = {
-  /**
-   * Available transforms.
-   */
-  transform: 'toggle';
-  /**
-   * Name of prop that transform should change. Must exist in TViewProps.
-   */
-  prop: keyof TViewProps;
-  /**
-   * Component prop setting default value.
-   */
-  defaultValueProp?: keyof TComponentProps;
-  /**
-   * Default value for prop when defaultValueProp is undefined.
-   */
-  defaultValue: boolean;
-  /**
-   * Callback as input into transform indicating change on input. Must exist in TViewProps.
-   */
-  onInput: keyof TViewProps;
-};
-
-export type IStateTransforms<TComponentProps, TViewProps> = (ITransform<TComponentProps, TViewProps>)[];
-
-export type IBaseStateComponentProps<TComponentProps, TViewProps> = IStateComponentProps<TComponentProps, TViewProps>;
+export interface IBaseStateOptions<TState> {
+  controlledProps: (keyof TState)[];
+}
 
-// TODO: One or the other of these solutions should work in TS. TypeScript is preventing use of
-//        keyof against generic types throughout this file.
-//        https://github.com/Microsoft/TypeScript/issues/13948
-// export type IBaseStateComponentState<TViewProps> = Partial<TViewProps>;
-export type IBaseStateComponentState<TViewProps> = {
-  // [key in keyof TViewProps]?: TViewProps[key];
-  // tslint:disable:no-any
-  [key in keyof TViewProps]?: any
-};
+export class BaseState<TProps, TState> extends React.Component<TProps, TState> {
+  private _controlledProps: (keyof TState)[];
 
-export class BaseStateComponent<TComponentProps, TViewProps> extends BaseComponent<
-  IBaseStateComponentProps<TComponentProps, TViewProps>,
-  IBaseStateComponentState<TViewProps>
-> {
-  constructor(
-    props: IBaseStateComponentProps<TComponentProps, TViewProps>,
-    transforms: IStateTransforms<TComponentProps, TViewProps>
-  ) {
+  constructor(props: TProps, options: Partial<IBaseStateOptions<TState>> = {}) {
     super(props);
+    this._controlledProps = options.controlledProps || [];
+  }
 
-    const stateObject: IBaseStateComponentState<TViewProps> = {};
+  public componentWillReceiveProps(newProps: TProps): void {
+    for (const propName of this._controlledProps) {
+      // tslint:disable-next-line:no-any
+      const controlledPropValue = (newProps as any)[propName];
+
+      if (controlledPropValue !== undefined && controlledPropValue !== this.state[propName]) {
+        this.setState({
+          [propName]: controlledPropValue
+          // tslint:disable-next-line:no-any
+        } as any);
+      }
+    }
```

Comment:
Hmm, is a filter on update approach more performant than a filter on read approach? I wonder if doing this on every render would be better than having a source of truth read function for components to use...

Reply:
Let's revisit this file in another PR.